### ISSUE TITLE
Validity checks needed for networks built or edited by API function calls

### ISSUE THREAD

**Issue by LRossman** [OpenWaterAnalytics/EPANET#361]
When a network is read from an input file certain validity checks are made to insure that the hydraulic solver will run correctly. These include making sure there are no orphan junctions with no connecting links, that there is at least one tank or reservoir in the network, that all pumps have either a pump curve or power rating assigned to them, etc. Error codes 223 - 233 are used to report the cause of a failed validity check.

When a network is built entirely from API function calls (or even when these calls are used to augment a network read in from a file) no such validity checks are made before the hydraulic solver is opened. If the solver tries to operate on an invalid network its behavior is either undefined or it returns a "cannot solve network hydraulic equations" error without indicating what the cause of the problem was.

One solution is to move the validity checks out of the input file processing code units and into the hydraulic solver's `openhyd()` function.

**Comment by michaeltryby** [comment]
Another possibility would be to associate the validity check with the API function calls and then rewrite the input file processing code to use the API instead of interacting with the data model directly.

**Comment by LRossman** [comment]
Some validity checks can't be made just within API net-builder function calls. An example would be a node created with `EN_addnode` that never gets included in a link thus creating an orphan node which causes the hydraulic solver to fail. In the interest of moving the project along I think it's best right now to simply move the validation code from the input file reader into its own function called from either `openhyd `or from `EN_openH`.

**Comment by michaeltryby** [comment]
@LRossman are we going to allow addition and deletion of model elements at any point in a simulation? Or only before the start of the simulation? We may need to define some sort of “mode” to determine when changes to the data model are allowed and when they aren’t.

**Comment by LRossman** [comment]
@michaeltryby I agree. The `OpenHflag` and `OpenQflag` variables can be used to prevent the following functions from being executed when either flag is `TRUE`:
`EN_setflowunits, EN_addnode, EN_deletenode, EN_addlink, EN_deletelink, EN_setlinknodes, and EN_setlinktype.`
This would also require adding another error message (or two).

In addition, it would probably make sense to prevent the `EN_settimeparam` function from executing when the solvers are active and operating in extended period mode (that is, somewhere in the middle of an EPS run). Error 109 ("cannot change time parameter when solver is active") already exists but it's not called anywhere in the code.

**Comment by samhatchett** [comment]
I know this is not the purpose of the issue above, but in my use case, I do want to change time parameters during a simulation. I explicitly set all of the steps, duration, and `H_TIME/Q_TIME` at every step. I must bypass epanet's notion of time to allow for continuous simulation because we don't know how long a solution will be valid until something happens in the physical system.

**Comment by LRossman** [comment]
OK @samhatchett , the idea of modifying `EN_settimeparam` is shelved.
